# PANDAS

FUNCTION INCLUDE ANALYZING , CLEANING AND MANIPULATING DATA WHICH ARE USED FOR WORKING WITH BIG DATASET 

PANDAS CAN CLEAN MESSY DATA SET , AND MAKE THEM REDABLE AND REVELENT.

In [19]:
import pandas as pd

In [46]:
# to read file

df = pd.read_csv("titanic.csv")

In [48]:
# to see top 5 data
# NaN are the missing data

df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [50]:
# to see last 5 data

df.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

In [52]:
# types of data

df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [61]:
# GIVES ALL THE DATA OF EVERY COLOUMN
# IN NUMERICAL DATA ONLY WE COULD USE (DESCRIBE) FUNCTION

df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

Here we can see that total value of age = 714 , but we have  total rows is 891 that means the rest values
in age are all null values (NaN).

In [65]:
# WHEN WE NEED SOME SPECIFIC DATASET FROM THE DATA

df[['Name' , 'Sex' , 'Ticket' , 'Cabin' , 'Embarked']]

Name     Sex  \
0                              Braund, Mr. Owen Harris    male   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2                               Heikkinen, Miss. Laina  female   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4                             Allen, Mr. William Henry    male   
..                                                 ...     ...   
886                              Montvila, Rev. Juozas    male   
887                       Graham, Miss. Margaret Edith  female   
888           Johnston, Miss. Catherine Helen "Carrie"  female   
889                              Behr, Mr. Karl Howell    male   
890                                Dooley, Mr. Patrick    male   

               Ticket Cabin Embarked  
0           A/5 21171   NaN        S  
1            PC 17599   C85        C  
2    STON/O2. 3101282   NaN        S  
3              113803  C123        S  
4              373450   NaN        S  
..                ...   ...      ...  
886            211536   NaN        S  
887            112053   B42        S  
888        W./C. 6607   NaN        S  
889            111369  C148        C  
890            370376   NaN        Q  

[891 rows x 5 columns]

In [84]:
# WE HAVE WRITTEN (OBJECT) BECAUSE WE WWANT TO KNOW WHICH ARE STRINGS

df.dtypes =='object'


PassengerId    False
Survived       False
Pclass         False
Name            True
Sex             True
Age            False
SibSp          False
Parch          False
Ticket          True
Fare           False
Cabin           True
Embarked        True
dtype: bool

In [82]:
df.dtypes 

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [86]:
df.dtypes == 'float64'

PassengerId    False
Survived       False
Pclass         False
Name           False
Sex            False
Age             True
SibSp          False
Parch          False
Ticket         False
Fare            True
Cabin          False
Embarked       False
dtype: bool

In [90]:
# MASKING for FLOAT64

df[df.dtypes[df.dtypes == 'float64'].index]

Age     Fare
0    22.0   7.2500
1    38.0  71.2833
2    26.0   7.9250
3    35.0  53.1000
4    35.0   8.0500
..    ...      ...
886  27.0  13.0000
887  19.0  30.0000
888   NaN  23.4500
889  26.0  30.0000
890  32.0   7.7500

[891 rows x 2 columns]

In [92]:
# MASKING for Int64

df[df.dtypes[df.dtypes == 'int64'].index]

PassengerId  Survived  Pclass  SibSp  Parch
0              1         0       3      1      0
1              2         1       1      1      0
2              3         1       3      0      0
3              4         1       1      1      0
4              5         0       3      0      0
..           ...       ...     ...    ...    ...
886          887         0       2      0      0
887          888         1       1      0      0
888          889         0       3      1      2
889          890         1       1      0      0
890          891         0       3      0      0

[891 rows x 5 columns]

In [94]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [96]:
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [100]:
df.tail(15)

PassengerId  Survived  Pclass  \
876          877         0       3   
877          878         0       3   
878          879         0       3   
879          880         1       1   
880          881         1       2   
881          882         0       3   
882          883         0       3   
883          884         0       2   
884          885         0       3   
885          886         0       3   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                              Name     Sex   Age  SibSp  \
876                  Gustafsson, Mr. Alfred Ossian    male  20.0      0   
877                           Petroff, Mr. Nedelio    male  19.0      0   
878                             Laleff, Mr. Kristo    male   NaN      0   
879  Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
880   Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
881                             Markun, Mr. Johann    male  33.0      0   
882                   Dahlberg, Miss. Gerda Ulrika  female  22.0      0   
883                  Banfield, Mr. Frederick James    male  28.0      0   
884                         Sutehall, Mr. Henry Jr    male  25.0      0   
885           Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                          Montvila, Rev. Juozas    male  27.0      0   
887                   Graham, Miss. Margaret Edith  female  19.0      0   
888       Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                          Behr, Mr. Karl Howell    male  26.0      0   
890                            Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
876      0              7534   9.8458   NaN        S  
877      0            349212   7.8958   NaN        S  
878      0            349217   7.8958   NaN        S  
879      1             11767  83.1583   C50        C  
880      1            230433  26.0000   NaN        S  
881      0            349257   7.8958   NaN        S  
882      0              7552  10.5167   NaN        S  
883      0  C.A./SOTON 34068  10.5000   NaN        S  
884      0   SOTON/OQ 392076   7.0500   NaN        S  
885      5            382652  29.1250   NaN        Q  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q

In [106]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [110]:
df['Ticket']

0             A/5 21171
1              PC 17599
2      STON/O2. 3101282
3                113803
4                373450
             ...       
886              211536
887              112053
888          W./C. 6607
889              111369
890              370376
Name: Ticket, Length: 891, dtype: object

# SLICING

In [114]:
df[['Ticket']][4:16]

Ticket
4      373450
5      330877
6       17463
7      349909
8      347742
9      237736
10    PP 9549
11     113783
12  A/5. 2151
13     347082
14     350406
15     248706

In [116]:
df[['Ticket']][4:16:2]

Ticket
4      373450
6       17463
8      347742
10    PP 9549
12  A/5. 2151
14     350406

In [124]:
df[['Ticket','Cabin']][4:16]

Ticket Cabin
4      373450   NaN
5      330877   NaN
6       17463   E46
7      349909   NaN
8      347742   NaN
9      237736   NaN
10    PP 9549    G6
11     113783  C103
12  A/5. 2151   NaN
13     347082   NaN
14     350406   NaN
15     248706   NaN

In [136]:
# CREATING A NEW COLOUMN WHERE WE ADD NEW COLOUMN

df['new_col']  = 45

In [138]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  new_col  
0        0         A/5 21171   7.2500   NaN        S       45  
1        0          PC 17599  71.2833   C85        C       45  
2        0  STON/O2. 3101282   7.9250   NaN        S       45  
3        0            113803  53.1000  C123        S       45  
4        0            373450   8.0500   NaN        S       45  
..     ...               ...      ...   ...      ...      ...  
886      0            211536  13.0000   NaN        S       45  
887      0            112053  30.0000   B42        S       45  
888      2        W./C. 6607  23.4500   NaN        S       45  
889      0            111369  30.0000  C148        C       45  
890      0            370376   7.7500   NaN        Q       45  

[891 rows x 13 columns]

In [154]:
# CREATING NEW COLOUMN AT SPECIFIED LOCATION

df.insert(loc = 3 , column = 'Food' , value = 7879)

ValueError: cannot insert Food, already exists

In [148]:
df

PassengerId  Survived  Pclass  Food  \
0              1         0       3     0   
1              2         1       1     0   
2              3         1       3     0   
3              4         1       1     0   
4              5         0       3     0   
..           ...       ...     ...   ...   
886          887         0       2     0   
887          888         1       1     0   
888          889         0       3     0   
889          890         1       1     0   
890          891         0       3     0   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  new_col  
0        0         A/5 21171   7.2500   NaN        S       45  
1        0          PC 17599  71.2833   C85        C       45  
2        0  STON/O2. 3101282   7.9250   NaN        S       45  
3        0            113803  53.1000  C123        S       45  
4        0            373450   8.0500   NaN        S       45  
..     ...               ...      ...   ...      ...      ...  
886      0            211536  13.0000   NaN        S       45  
887      0            112053  30.0000   B42        S       45  
888      2        W./C. 6607  23.4500   NaN        S       45  
889      0            111369  30.0000  C148        C       45  
890      0            370376   7.7500   NaN        Q       45  

[891 rows x 14 columns]

In [156]:
df.insert(loc = 3 , column = 'Equipment' , value = 'oxygen cylinder')

In [158]:
df

PassengerId  Survived  Pclass        Equipment  Food  \
0              1         0       3  oxygen cylinder     0   
1              2         1       1  oxygen cylinder     0   
2              3         1       3  oxygen cylinder     0   
3              4         1       1  oxygen cylinder     0   
4              5         0       3  oxygen cylinder     0   
..           ...       ...     ...              ...   ...   
886          887         0       2  oxygen cylinder     0   
887          888         1       1  oxygen cylinder     0   
888          889         0       3  oxygen cylinder     0   
889          890         1       1  oxygen cylinder     0   
890          891         0       3  oxygen cylinder     0   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  new_col  
0        0         A/5 21171   7.2500   NaN        S       45  
1        0          PC 17599  71.2833   C85        C       45  
2        0  STON/O2. 3101282   7.9250   NaN        S       45  
3        0            113803  53.1000  C123        S       45  
4        0            373450   8.0500   NaN        S       45  
..     ...               ...      ...   ...      ...      ...  
886      0            211536  13.0000   NaN        S       45  
887      0            112053  30.0000   B42        S       45  
888      2        W./C. 6607  23.4500   NaN        S       45  
889      0            111369  30.0000  C148        C       45  
890      0            370376   7.7500   NaN        Q       45  

[891 rows x 15 columns]